flang/lib/Lower/Bridge.cpp

```diff
@@ -1805,6 +1805,10 @@ class FirConverter : public Fortran::lower::AbstractConverter {
                 // Assume lhs is not polymorphic for now given TODO above,
                 // otherwise, the read would is conservative and returns
                 // BoxValue for derived types.
+                // FIXME: Why does this return the value of the lhs variable
+                // instead of a reference, the lhs variable's location? For an
+                // assignment, the value is going to be dead shortly and doesn't
+                // seem useful.
                 return Fortran::lower::genMutableBoxRead(
```

Comment:
`genMutableBoxRead` name may mislead you here. It is reading the address of allocatable/pointer, not the value.

What the code is doing here is dealing with the case where the LHS is an allocatable (say a  `float**`). Once the box is allocated according to the needs, we want to get the `float*` pointer so that we can assign to it. That is what `genMutableBoxRead` is doing, it is doing a single dereference of `float**` into a `float*`, it is not reading into a value.

Reply:
+1